konfera/models/ticket.py

```diff
@@ -44,7 +45,14 @@ def discount_calculator(self):
             return self.type.price * self.discount_code.discount / 100
         return 0
 
+    def clean(self):
+        if self.discount_code and self.discount_code.ticket_type != self.type:
+            raise ValidationError(
+                {'discount_code': _('Discount code is not applicable for the selected ticket type')})
+        super(Ticket, self).clean()
+
     def save(self, *args, **kwargs):
+        self.clean()
```

Comment:
This makes it so the clean logic is run when someone tries to save a ticket object programmatically.